### PR TITLE
notehead: refactor any in glyphProps

### PR DESCRIPTION
**Pull request by rvilarl**
As discussed in #1415 first step removing any from tables.ts

No visual differences.

### CODE REVIEW

**Comment by rvilarl** [comment]
@ronyeh could you review?

File: src/glyph.ts [hunk 1]
```diff
@@ -7,34 +7,33 @@ import { Element } from './element';
...
-  width: number;
-  text: string;
-  tabnote_stem_down_extension: number;
-  tabnote_stem_up_extension: number;
-  beam_count: number;
-  shift_y: number;
-  getWidth(a?: number): number;
-  getMetrics(): GlyphMetrics;
+  code_head?: string;
+  leger_code_head?: string;
```
Comment (sschmidTU) [inline src/glyph.ts:16]: it's `ledger` by the way, not `ledger. like `ledger note`.

Comment (rvilarl) [reply src/glyph.ts:16]: I also thought about this one, and I found:

https://en.m.wikipedia.org/wiki/Ledger_line

Perhaps we should unify.

**Comment by rvilarl** [comment]
I also have a proposal:
this.glyph and this.code (variables used to store the value returned by glyphProps) are confusing. I would suggest to rename them to this.glyphProps

File: src/glyph.ts (hunk 1 above)
Comment (AaronDavidNewman) [reply src/glyph.ts:16]: @sschmidTU Beginner music books in the US use ledger.  So Hal Leonard does not agree :)  But I went with leger b/c that's what Smufl uses, and I feel like we're a Smufl shop.

**Comment by rvilarl** [comment]
> I also have a proposal: this.glyph and this.code (variables used to store the value returned by getGlyphProps) are confusing. I would suggest to rename them to this.glyphProps

Done

**Comment by rvilarl** [comment]
Doing the change I realised that `ClefNote` should inherit from `GlyphNote`

**Comment by rvilarl** [comment]
No vvisual differences and ready for review

File: src/glyph.ts (hunk 1 above)
Comment (0xfe) [reply src/glyph.ts:16]: I prefer using `ledger` in this codebase (mainly to be consistent.)

Comment (rvilarl) [reply src/glyph.ts:16]: Ok I understand everything ledger except for the Smufl glyph IDs.

**Comment by rvilarl** [comment]
@0xfe leger topic noted. Anything else? This is ready for review.

File: src/glyph.ts (hunk 1 above)
Comment (rvilarl) [reply src/glyph.ts]: Done

File: src/multimeasurerest.ts [hunk 2]
```diff
@@ -227,7 +227,7 @@ export class MultiMeasureRest extends Element {
       x += glyphs[2].width + spacing;
     }
     for (let i = 0; i < n1; ++i) {
-      Glyph.renderGlyph(ctx, x, yTop, rest_scale, rest.glyph_code);
+      if (rest.glyph_code) Glyph.renderGlyph(ctx, x, yTop, rest_scale, rest.glyph_code);
```
Comment (0xfe) [inline src/multimeasurerest.ts:230]: This looks like a potentially silent bug. Shouldn't there always be a `glyph_code` if we're here?

**Review by 0xfe — COMMENTED** [review]
Hi Rodrigo -- I like that ClefNote is not based on GlyphNote, however I'm not sure that this change to eliminate `any` is worth it.

We've replaced a lot of required struct members to optional, and now it becomes very difficult to reason about in code because it's hard to know if an instance of that struct will have a member set or not, so there's a littering of `if (glyphProps.foo) { ... }`, which is either unnecessary code, or creates silent failures.

I think it makes the codebase more complicated.

**Comment by rvilarl** [comment]
> Hi Rodrigo -- I like that ClefNote is not based on GlyphNote, 

Ok I can revert this one.

> however I'm not sure that this change to eliminate `any` is worth it.
> We've replaced a lot of required struct members to optional...

I believe that elminating any is what reveals silent failures. I have only made optional those members which where optionally defined. If they are required, we should make them required and make sure that they are defined which is today, I believe, not the case.

File: src/multimeasurerest.ts (hunk 2 above)
Comment (rvilarl) [reply src/multimeasurerest.ts:230]: Slash notes do not define a `code_head` this is why `glyph_code` is also optional. I could define `code_head` for slashes as 'N/A' or '' and make the attribute not optional. Would you prefer this approach?

Comment (rvilarl) [reply src/multimeasurerest.ts]: code_head no longer optional.

**Comment by rvilarl** [comment]
ClefNote inherits from Note
@0xfe ready to review

**Comment by 0xfe** [comment]
On Sat., Jul. 9, 2022, 11:43 Rodrigo Vilar, ***@***.***>
wrote:

> Hi Rodrigo -- I like that ClefNote is not based on GlyphNote,
>
> Ok I can revert this one.
>

Sorry typo, I meant _now_ not _not_ -- I like this change. Should prob be a
separate PR.

> however I'm not sure that this change to eliminate any is worth it.
> We've replaced a lot of required struct members to optional...
>
> I believe that elminating any is what reveals silent failures.
>

Removing any this way actually masks them, not reveals them. With the way
it is right now (i.e., with any), at least incorrect code will fail loudly.
With these changes, you'll need to check for the existence of each optional
anytime the struct is used, and then do something specific about it -- the
easiest is to ignore the paths, which causes the silent failures.

Does that make sense?


ave only made optional those members which where optionally defined. If

**Comment by rvilarl** [comment]
>> however I'm not sure that this change to eliminate any is worth it.
>> We've replaced a lot of required struct members to optional...
>>
>> I believe that elminating any is what reveals silent failures.
>>
> Removing any this way actually masks them, not reveals them. With the way
> it is right now (i.e., with any), at least incorrect code will fail loudly.
> With these changes, you'll need to check for the existence of each optional
> anytime the struct is used, and then do something specific about it -- the
> easiest is to ignore the paths, which causes the silent failures.
> 
> Does that make sense?

@0xfe I think that we are consideering reveal in two different scenarios. I am considering compile-time (this is how I discovered that stem was a boolean) and you are considering run-time. If you want to fail them as before I can use the `as` or the `!` operators rather than checking with `if`. Please note that this PR also resolves the overlap of `GlyphProps` and `Glyph` which is very confusing.

**Comment by 0xfe** [comment]
Yes, I understand that removing `any` gives us compile-time checking for free -- I'm not disputing that. I'm saying that the tradeoff here is creating lots of optional fields, that requires us to scatter run-time checks (with `if` or `!`) across the codebase which can mask errors, which IMO is worse.

(Right now the `any` will at least fail loudly at runtime if we got something wrong -- so long as we have good test coverage, we can keep the code robust.)

So if we want to eliminate `any`, we should think about this tradeoff. All the optionals here signal to me that there's some hidden substructures that we're not seeing.  E.g., is there a pattern where we can break up `glyphProps` into different interfaces or something?

**Comment by rvilarl** [comment]
I believe that there are two options:
1. separate `durationCodes` in two arrays one with the `common` elements and make it of type `GlyphProps` rather than `Partial<GlyphProps>`
2.  fill all the defaults in `getGlyphProps`and leave the `durationCodes` as `Partial<GlyphProps>`

@0xfe what do you think?

**Comment by rvilarl** [comment]
Closed to focus on #1409 in order to release 4.1

**Pull request closed without merging**